src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AdminControllerTests.cs

```diff
@@ -0,0 +1,218 @@
+            var exception = Assert.Throws<InvalidOperationException>(act);
+            Assert.Equal(exception.Message, ExceptionsMessages.TestAlreadyRun);
+        }
+
+        [Fact]
+        public void StartTest_PreviousTestIsNotDone_ThrowsException()
+        {
+            // Arrange
+            var test = GetTest(TestStatus.NotStarted);
+            var previousTest = GetTest(TestStatus.Started);
+            this._adminServiceMock.Setup(x => x.GetTestById(It.IsAny<int>())).Returns(test);
+            this._adminServiceMock.Setup(x => x.GetPreviousTest(It.IsAny<int>())).Returns(previousTest);
+            using var controller = new AdminController(this._adminServiceMock.Object, this._testServiceMock.Object);
+
+            // Act
+            Func<ActionResult> act = () => controller.StartTest(test.Id, "20");
+
+            // Assert
+            var exception = Assert.Throws<InvalidOperationException>(act);
+            Assert.Equal(exception.Message, ExceptionsMessages.PreviousTestIsNotDone);
+        }
+
+        [Fact]
+        public void StartTest_TestIsNotStartedAndPreviousOneIsNotDone_ReturnsARedirectAndUpdateTestDates()
+        {
+            // Arrange
+            var minutes = "20";
+            var test = GetTest(TestStatus.NotStarted);
+            var previousTest = GetTest(TestStatus.Ended);
+            this._adminServiceMock.Setup(x => x.GetTestById(It.IsAny<int>())).Returns(test);
+            this._adminServiceMock.Setup(x => x.GetPreviousTest(It.IsAny<int>())).Returns(previousTest);
+            using var controller = new AdminController(this._adminServiceMock.Object, this._testServiceMock.Object);
+
+            // Act
+            var result = controller.StartTest(test.Id, minutes);
+
+            // Assert
+            this._adminServiceMock.Verify(x => x.UpdateTestDates(test.Id, minutes), Times.Once);
+            var redirectToActionResult = Assert.IsType<RedirectToActionResult>(result);
+            Assert.Equal("Index", redirectToActionResult.ActionName);
+        }
+
+        [Fact]
+        public void CalculateResults_IncorrectWeekNumber_ReturnsBadRequestResult()
+        {
+            // Arrange
+            var incorrectWeekNumber = -1;
+            using var controller = new AdminController(this._adminServiceMock.Object, this._testServiceMock.Object);
+
+            // Act
+            var result = controller.CalculateResults(incorrectWeekNumber);
+
+            // Assert
+            Assert.IsType<BadRequestObjectResult>(result);
+        }
+
+        [Fact]
+        public void AdminController_ShouldHaveCorrectRole()
+        {
+            using var controller = new AdminController(this._adminServiceMock.Object, this._testServiceMock.Object);
+            var authorizeAttribute = GetAuthorizeAttribute(controller);
+
+            Assert.NotNull(authorizeAttribute);
+            Assert.NotNull(authorizeAttribute.Roles);
+            Assert.Equal("Admin", authorizeAttribute.Roles);
```

Comment:
we should ensure that every method in _AdminController_ require Admin privileges 
because you cand o sth like that:
```
[Authorize]
    public class WidgetController : Controller
    {
        // GET: Widget
        [AllowAnonymous]
        public ActionResult Index()
        {
            return View();
        }
}
```
and don't want that situation ;)